src/AddAtomicMutex.cpp

```diff
@@ -0,0 +1,340 @@
+        store_names.push(op->name);
+    }
+};
+
+/** Clear out the Atomic node's mutex usages if it doesn't need one. */
+class RemoveUnnecessaryMutexUse : public IRMutator {
+public:
+    set<string> remove_mutex_lock_names;
+
+protected:
+    using IRMutator::visit;
+
+    Stmt visit(const Atomic *op) override {
+        // Collect the names of all Store nodes inside.
+        CollectStoreNames collector;
+        op->body.accept(&collector);
+        // Search for let bindings that access the producers.
+        FindAtomicLetBindings finder(collector.store_names);
+        op->body.accept(&finder);
+        if (finder.found) {
+            // Can't remove mutex lock. Leave the Stmt as is.
+            return IRMutator::visit(op);
+        } else {
+            remove_mutex_lock_names.insert(op->mutex_name);
+            Stmt body = mutate(op->body);
+            return Atomic::make(op->producer_name,
+                                string(),
+                                std::move(body));
+        }
+    }
+};
+
+/** Find Store inside an Atomic that matches the provided store_names. */
+class FindStoreInAtomicMutex : public IRGraphVisitor {
+public:
+    using IRGraphVisitor::visit;
+
+    FindStoreInAtomicMutex(const std::set<std::string> &store_names)
+        : store_names(store_names) {}
+
+    bool found = false;
+    std::string producer_name;
+    std::string mutex_name;
+protected:
+    void visit(const Atomic *op) override {
+        if (!found && !op->mutex_name.empty()) {
+            ScopedValue<bool> old_in_atomic_mutex(in_atomic_mutex, true);
+            include(op->body);
+            if (found) {
+                // We found a Store inside Atomic with matching name,
+                // record the mutex information.
+                producer_name = op->producer_name;
+                mutex_name = op->mutex_name;
+            }
+        } else {
+            include(op->body);
+        }
+    }
+
+    void visit(const Store *op) override {
+        if (in_atomic_mutex) {
+            if (store_names.find(op->name) != store_names.end()) {
+                found = true;
+            }
+        }
+        IRGraphVisitor::visit(op);
+    }
+
+    bool in_atomic_mutex = false;
+    const std::set<std::string> &store_names;
+};
+
+/** Find Store inside of an Atomic node and return their indices. */
+class FindStoreIndex : public IRGraphVisitor {
+public:
+    Expr index;
+protected:
+    using IRGraphVisitor::visit;
+
+    void visit(const Store *op) override {
+        // Ideally we want to insert equal() checks here for different stores,
+        // but the indices of them actually are different in the case of tuples,
+        // since they usually refer to the strides/min/extents of their own tuple
+        // buffers. However, different elements in a tuple would have the same
+        // strides/min/extents so we are fine.
+        if (index.defined()) {
+            return;
+        }
+        index = op->index;
```

Comment:
This lifts the index out of any containing LetStmts which may define variables it relies on. Add a LetStmt visitor:

```
void visit(const LetStmt *op) override {
  IRGraphVisitor::visit(op);
  if (index.defined()) {
    if (expr_uses_var(index, op->name)) {
      index = Let::make(op->name, op->value, index);
    }
  }
}
```

Reply:
Done.